File: teamscale-maven-plugin/src/main/java/com/teamscale/maven/tia/TiaCoverageConvertMojo.java
package com.teamscale.maven.tia;

import com.google.common.base.Strings;
import com.teamscale.jacoco.agent.options.AgentOptionParseException;
import com.teamscale.jacoco.agent.options.ClasspathUtils;
import com.teamscale.jacoco.agent.options.FilePatternResolver;
import org.apache.maven.execution.MavenSession;
import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.apache.maven.plugins.annotations.ResolutionScope;
import org.apache.maven.project.MavenProject;
import shadow.com.teamscale.client.TestDetails;
import shadow.com.teamscale.report.EDuplicateClassFileBehavior;
import shadow.com.teamscale.report.ReportUtils;
import shadow.com.teamscale.report.testwise.ETestArtifactFormat;
import shadow.com.teamscale.report.testwise.TestwiseCoverageReportWriter;
import shadow.com.teamscale.report.testwise.jacoco.JaCoCoTestwiseReportGenerator;
import shadow.com.teamscale.report.testwise.model.TestExecution;
import shadow.com.teamscale.report.testwise.model.factory.TestInfoFactory;
import shadow.com.teamscale.report.util.ClasspathWildcardIncludeFilter;
import shadow.com.teamscale.report.util.CommandLineLogger;
import shadow.com.teamscale.report.util.ILogger;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

/**
 * Batch converts all created .exec file reports into a testwise coverage
 * report.
 */
@Mojo(name = "testwise-coverage-converter", defaultPhase = LifecyclePhase.VERIFY, requiresDependencyResolution = ResolutionScope.RUNTIME)
public class TiaCoverageConvertMojo extends AbstractMojo {
	/**
	 * Wildcard include patterns to apply during JaCoCo's traversal of class files.
	 */
	@Parameter(defaultValue = "**")
	public String[] includes;
	/**
	 * Wildcard exclude patterns to apply during JaCoCo's traversal of class files.
	 */
	@Parameter()
	public String[] excludes;

	/**
	 * After how many tests the testwise coverage should be split into multiple
	 * reports (Default is 5000).
	 */
	@Parameter(defaultValue = "5000")
	public int splitAfter;

	/**
	 * The project build directory (usually: {@code ./target}). Provided
	 * automatically by Maven.
	 */
	@Parameter(defaultValue = "${project.build.directory}")
	public String projectBuildDir;

	/**
	 * The output directory of the testwise coverage reports.
	 */
	@Parameter()
	public String outputFolder;

	/**
	 * The running Maven session. Provided automatically by Maven.
	 */
	@Parameter(defaultValue = "${session}")
	public MavenSession session;
	private final ILogger logger = new CommandLineLogger();

	@Override
	public void execute() throws MojoFailureException {

		List<File> reportFileDirectories = new ArrayList<>();
		reportFileDirectories.add(Paths.get(projectBuildDir, "tia").toAbsolutePath().resolve("reports").toFile());
		List<File> classFileDirectories;
		if (Strings.isNullOrEmpty(outputFolder)) {
			outputFolder = Paths.get(projectBuildDir, "tia", "reports").toString();
		}
		try {
			Files.createDirectories(Paths.get(outputFolder));
			classFileDirectories = getClassDirectoriesOrZips(projectBuildDir);
			findSubprojectReportAndClassDirectories(reportFileDirectories, classFileDirectories);
		} catch (IOException | AgentOptionParseException e) {
			logger.error("Could not create testwise report generator. Aborting.", e);
			throw new MojoFailureException(e);
		}
		logger.info("Generating the testwise coverage report");
		JaCoCoTestwiseReportGenerator generator = createJaCoCoTestwiseReportGenerator(classFileDirectories);
		TestInfoFactory testInfoFactory = createTestInfoFactory(reportFileDirectories);
		List<File> jacocoExecutionDataList = ReportUtils.listFiles(ETestArtifactFormat.JACOCO, reportFileDirectories);
		String reportFilePath = Paths.get(outputFolder, "testwise-coverage.json").toString();

		try (TestwiseCoverageReportWriter coverageWriter = new TestwiseCoverageReportWriter(testInfoFactory,
				new File(reportFilePath), splitAfter)) {
			for (File executionDataFile : jacocoExecutionDataList) {
				logger.info("Writing execution data for file: " + executionDataFile.getName());
				generator.convertAndConsume(executionDataFile, coverageWriter);
			}
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	private void findSubprojectReportAndClassDirectories(List<File> reportFiles,
														 List<File> classFiles) throws AgentOptionParseException {

		for (MavenProject subProject : session.getTopLevelProject().getCollectedProjects()) {
			String subprojectBuildDirectory = subProject.getBuild().getDirectory();
			reportFiles.add(Paths.get(subprojectBuildDirectory, "tia").toAbsolutePath().resolve("reports")
					.toFile());
			classFiles.addAll(getClassDirectoriesOrZips(subprojectBuildDirectory));
		}
	}

	private TestInfoFactory createTestInfoFactory(List<File> reportFiles) throws MojoFailureException {
		try {
			List<TestDetails> testDetails = ReportUtils.readObjects(ETestArtifactFormat.TEST_LIST, TestDetails[].class,
					reportFiles);
			List<TestExecution> testExecutions = ReportUtils.readObjects(ETestArtifactFormat.TEST_EXECUTION,
					TestExecution[].class, reportFiles);
			logger.info("Writing report with " + testDetails.size() + " Details/" + testExecutions.size() + " Results");
			return new TestInfoFactory(testDetails, testExecutions);
		} catch (IOException e) {
			logger.error("Could not read test details from reports. Aborting.", e);
			throw new MojoFailureException(e);
		}
	}

	private JaCoCoTestwiseReportGenerator createJaCoCoTestwiseReportGenerator(List<File> classFiles) {
		String includes = null;
		if (this.includes != null) {
			includes = String.join(":", this.includes);
		}
		String excludes = null;
		if (this.excludes != null) {
			excludes = String.join(":", this.excludes);
		}
		return new JaCoCoTestwiseReportGenerator(classFiles,
				new ClasspathWildcardIncludeFilter(includes, excludes), EDuplicateClassFileBehavior.WARN, logger);
	}

	private List<File> getClassDirectoriesOrZips(String projectBuildDir) throws AgentOptionParseException {
		List<String> classDirectoriesOrZips = new ArrayList<>();
		classDirectoriesOrZips.add(projectBuildDir);
		return ClasspathUtils.resolveClasspathTextFiles("classes", new FilePatternResolver(new CommandLineLogger()),
				classDirectoriesOrZips);
	}
}

File: teamscale-maven-plugin/src/main/kotlin/com/teamscale/maven/tia/TiaCoverageConvertMojo.kt
package com.teamscale.maven.tia

import com.teamscale.jacoco.agent.options.AgentOptionParseException
import com.teamscale.jacoco.agent.options.ClasspathUtils
import com.teamscale.jacoco.agent.options.FilePatternResolver
import org.apache.maven.execution.MavenSession
import org.apache.maven.plugin.AbstractMojo
import org.apache.maven.plugin.MojoFailureException
import org.apache.maven.plugins.annotations.LifecyclePhase
import org.apache.maven.plugins.annotations.Mojo
import org.apache.maven.plugins.annotations.Parameter
import org.apache.maven.plugins.annotations.ResolutionScope
import shadow.com.teamscale.client.TestDetails
import shadow.com.teamscale.report.EDuplicateClassFileBehavior
import shadow.com.teamscale.report.ReportUtils
import shadow.com.teamscale.report.testwise.ETestArtifactFormat
import shadow.com.teamscale.report.testwise.TestwiseCoverageReportWriter
import shadow.com.teamscale.report.testwise.jacoco.JaCoCoTestwiseReportGenerator
import shadow.com.teamscale.report.testwise.model.TestExecution
import shadow.com.teamscale.report.testwise.model.factory.TestInfoFactory
import shadow.com.teamscale.report.util.ClasspathWildcardIncludeFilter
import shadow.com.teamscale.report.util.CommandLineLogger
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

/**
 * Batch converts all created .exec file reports into a testwise coverage
 * report.
 */
@Mojo(
	name = "testwise-coverage-converter",
	defaultPhase = LifecyclePhase.VERIFY,
	requiresDependencyResolution = ResolutionScope.RUNTIME
)
class TiaCoverageConvertMojo : AbstractMojo() {
	/**
	 * Wildcard include patterns to apply during JaCoCo's traversal of class files.
	 */
	@Parameter(defaultValue = "**")
	lateinit var includes: Array<String>

	/**
	 * Wildcard exclude patterns to apply during JaCoCo's traversal of class files.
	 */
	@Parameter
	lateinit var excludes: Array<String>

	/**
	 * After how many tests the testwise coverage should be split into multiple
	 * reports (Default is 5000).
	 */
	@Parameter(defaultValue = "5000")
	var splitAfter: Int = 5000

	/**
	 * The project build directory (usually: `./target`). Provided
	 * automatically by Maven.
	 */
	@Parameter(defaultValue = "\${project.build.directory}")
	lateinit var projectBuildDir: String

	/**
	 * The output directory of the testwise coverage reports.
	 */
	@Parameter
	lateinit var outputFolder: String

	/**
	 * The running Maven session. Provided automatically by Maven.
	 */
	@Parameter(defaultValue = "\${session}")
	lateinit var session: MavenSession
	private val logger = CommandLineLogger()

	@Throws(MojoFailureException::class)
	override fun execute() {
		val reportFileDirectories = mutableListOf<File>()
		reportFileDirectories.add(Paths.get(projectBuildDir, "tia").toAbsolutePath().resolve("reports").toFile())
		if (outputFolder.isBlank()) {
			outputFolder = Paths.get(projectBuildDir, "tia", "reports").toString()
		}
		val classFileDirectories: MutableList<File>
		try {
			Files.createDirectories(Paths.get(outputFolder))
			classFileDirectories = getClassDirectoriesOrZips(projectBuildDir)
			findSubprojectReportAndClassDirectories(reportFileDirectories, classFileDirectories)
		} catch (e: IOException) {
			logger.error("Could not create testwise report generator. Aborting.", e)
			throw MojoFailureException(e)
		} catch (e: AgentOptionParseException) {
			logger.error("Could not create testwise report generator. Aborting.", e)
			throw MojoFailureException(e)
		}
		logger.info("Generating the testwise coverage report")
		val generator = createJaCoCoTestwiseReportGenerator(classFileDirectories)
		val testInfoFactory = createTestInfoFactory(reportFileDirectories)
		val jacocoExecutionDataList = ReportUtils.listFiles(ETestArtifactFormat.JACOCO, reportFileDirectories)
		val reportFilePath = Paths.get(outputFolder, "testwise-coverage.json").toString()

		try {
			TestwiseCoverageReportWriter(
				testInfoFactory, File(reportFilePath), splitAfter
			).use { coverageWriter ->
				jacocoExecutionDataList.forEach { executionDataFile ->
					logger.info("Writing execution data for file: ${executionDataFile.name}")
					generator.convertAndConsume(executionDataFile, coverageWriter)
				}
			}
		} catch (e: IOException) {
			throw RuntimeException(e)
		}
	}

	@Throws(AgentOptionParseException::class)
	private fun findSubprojectReportAndClassDirectories(
		reportFiles: MutableList<File>,
		classFiles: MutableList<File>
	) {
		session.topLevelProject.collectedProjects.forEach { subProject ->
			val subprojectBuildDirectory = subProject.build.directory
			reportFiles.add(
				Paths.get(subprojectBuildDirectory, "tia").toAbsolutePath().resolve("reports").toFile()
			)
			classFiles.addAll(getClassDirectoriesOrZips(subprojectBuildDirectory))
		}
	}

	@Throws(MojoFailureException::class)
	private fun createTestInfoFactory(reportFiles: List<File>): TestInfoFactory {
		try {
			val testDetails = ReportUtils.readObjects(
				ETestArtifactFormat.TEST_LIST,
				Array<TestDetails>::class.java,
				reportFiles
			)
			val testExecutions = ReportUtils.readObjects(
				ETestArtifactFormat.TEST_EXECUTION,
				Array<TestExecution>::class.java,
				reportFiles
			)
			logger.info("Writing report with ${testDetails.size} Details/${testExecutions.size} Results")
			return TestInfoFactory(testDetails, testExecutions)
		} catch (e: IOException) {
			logger.error("Could not read test details from reports. Aborting.", e)
			throw MojoFailureException(e)
		}
	}

	private fun createJaCoCoTestwiseReportGenerator(classFiles: List<File>) =
		JaCoCoTestwiseReportGenerator(
			classFiles,
			ClasspathWildcardIncludeFilter(
				includes.joinToString(":"),
				excludes.joinToString(":")
			), EDuplicateClassFileBehavior.WARN, logger
		)

	@Throws(AgentOptionParseException::class)
	private fun getClassDirectoriesOrZips(projectBuildDir: String) =
		ClasspathUtils.resolveClasspathTextFiles(
			"classes",
			FilePatternResolver(CommandLineLogger()),
			listOf(projectBuildDir)
		)
}
